onmt/dynamic/iterator.py

```diff
@@ -0,0 +1,127 @@
+        return dataset
+
+
+class DynamicDatasetIter(object):
+    """Yield data from multiply sharded plain text files."""
+
+    def __init__(self, corpora_shards, transforms, fields, opts, is_train):
+        self.corpora_shards = corpora_shards
+        self.transforms = transforms
+        self.fields = fields
+        self.corpora_info = opts.data
+        self.is_train = is_train
+
+        self.batch_size = opts.batch_size if is_train \
+            else opts.valid_batch_size
+        self.batch_size_fn = max_tok_len \
+            if opts.batch_type == "tokens" else None
+        if opts.batch_size_multiple is not None:
+            self.batch_size_multiple = opts.batch_size_multiple
+        else:
+            self.batch_size_multiple = 8 if opts.model_dtype == "fp16" else 1
+        self.device = 'cpu'
+        self.sort_key = str2sortkey[opts.data_type]
+        self.bucket_size = opts.bucket_size
+        self.pool_factor = opts.pool_factor
+
+    def _init_datasets(self):
+        self.datasets_iterables = build_sharded_corpora_iters(
+            self.corpora_shards, self.transforms,
+            self.corpora_info, self.is_train)
+        self.dataset_adapter = DatasetAdapter(self.fields)
+        self.datasets_weights = {
+            ds_name: int(self.corpora_info[ds_name]['weight'])
+            for ds_name in self.datasets_iterables.keys()
+        }
+        self.datasets_iters = {
+            ds_name: iter(generator)
+            for ds_name, generator in self.datasets_iterables.items()
+        }
+
+    def _iter_datasets(self):
+        """Yield a bucket of examples from weighted datasets."""
+        for ds_name, ds_weight in self.datasets_weights.items():
+            for _ in range(ds_weight):
+                yield self.datasets_iters[ds_name]
+
+    def _iter_examples(self):
+        for iterator in cycle(self._iter_datasets()):
+            yield next(iterator)
+
+    def _bucketing(self):
+        buckets = torchtext_batch(
+            self._iter_examples(),
+            batch_size=self.bucket_size,
+            batch_size_fn=None)
+        yield from buckets
+
+    def __iter__(self):
+        self._init_datasets()
+        while True:
+            for bucket in self._bucketing():
+                dataset = self.dataset_adapter(bucket)
+                train_iter = OrderedIterator(
+                    dataset,
+                    self.batch_size,
+                    pool_factor=self.pool_factor,
+                    batch_size_fn=self.batch_size_fn,
+                    batch_size_multiple=self.batch_size_multiple,
+                    device=self.device,
+                    train=self.is_train,
+                    sort=False,
+                    sort_within_batch=True,
+                    sort_key=self.sort_key,
+                    repeat=False,
+                )
+                for batch in train_iter:
+                    # if mb_callback is not None:
```

Comment:
?

Reply:
The `mb_callback` can be used to log statistic info about transform operation which I haven't implemented. Will think about it when complete transform.